# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/HttpChannelParent.cpp
# Commit: cad764eb15a4
# Full Hash: cad764eb15a4a42cfdb2468ce762029130c0b9fc
# Author: Junior Hsu <juhsu@mozilla.com>
# Date: 2020-06-30 10:00:42
# Regressor Bug: 1633935
# File Overlap Count: 1
# Description:
#   Bug 1633935 - P12 top-level IPC stream filter should happen before OnStartRequest, r=mayhemer,necko-reviewers
#   
#   The WebRequest will suspend the channel in nsHttpChannel::PrepareToConnect(), apply the stream filter, send the main thread IPC  and resume.
#   That is, PHttpChannel::AttachStreamFilter should be handled in Child before OnStartRequest goes to listener.
#   
# ==============================================================================

diff -r f75cc7d5f40e -r cad764eb15a4 netwerk/protocol/http/HttpChannelParent.cpp
--- a/netwerk/protocol/http/HttpChannelParent.cpp	Tue Jun 30 00:34:24 2020 +0000
+++ b/netwerk/protocol/http/HttpChannelParent.cpp	Tue Jun 30 00:34:32 2020 +0000
@@ -83,7 +83,8 @@
       mCacheNeedFlowControlInitialized(false),
       mNeedFlowControl(true),
       mSuspendedForFlowControl(false),
-      mAfterOnStartRequestBegun(false) {
+      mAfterOnStartRequestBegun(false),
+      mStreamFilterAttached(false) {
   LOG(("Creating HttpChannelParent [this=%p]\n", this));
 
   // Ensure gHttpHandler is initialized: we need the atom table up and running.
@@ -1541,7 +1542,8 @@
   // Bug 1645901: Currently Set-Cookie is passed to child process on main
   // thread, which is racy with PBackground. We should have a way to set cookie
   // in child for Set-Cookie response header.
-  args.shouldWaitForOnStartRequestSent() = isDocument || hasSetCookie;
+  args.shouldWaitForOnStartRequestSent() =
+      isDocument || hasSetCookie || mStreamFilterAttached;
 
   rv = NS_OK;
 
@@ -2663,5 +2665,12 @@
   mOverrideReferrerInfo = aReferrerInfo;
 }
 
+bool HttpChannelParent::AttachStreamFilter(
+    Endpoint<extensions::PStreamFilterParent>&& aEndpoint) {
+  MOZ_ASSERT(!mAfterOnStartRequestBegun);
+  mStreamFilterAttached = true;
+  return SendAttachStreamFilter(std::move(aEndpoint));
+}
+
 }  // namespace net
 }  // namespace mozilla